xlators/cluster/afr/src/afr-open.c

```diff
@@ -271,18 +267,145 @@ afr_fd_ctx_need_open(fd_t *fd, xlator_t *this, unsigned char *need_open)
+        if (ret)
+            gf_msg(this->name, GF_LOG_ERROR, -ret, AFR_MSG_DICT_GET_FAILED,
+                   "Failed to get value of strict-locks setting");
+    }
+
+    syncbarrier_wake(&local->barrier);
+    return 0;
+}
+
+int
+afr_do_fix_open(void *opaque)
+{
+    call_frame_t *frame = NULL;
+    xlator_t *this = NULL;
+    afr_private_t *priv = NULL;
+    afr_local_t *local = NULL;
+    dict_t *xdata = NULL;
+    gf_boolean_t locks_held = _gf_false;
+    unsigned char *success_replies = NULL;
+    unsigned char *need_open = NULL;
+    int i = 0;
+    int need_open_count = 0;
+    struct gf_flock flock = {
+        0,
+    };
+
+    frame = (call_frame_t *)opaque;
+    local = frame->local;
+    this = frame->this;
+    priv = this->private;
+
+    local->cont.lk.getlk_rsp1 = GF_CALLOC(sizeof(*local->cont.lk.getlk_rsp1),
+                                          priv->child_count, gf_afr_mt_gf_lock);
+
+    flock.l_type = F_WRLCK;
+    afr_set_lk_owner(frame, this, frame->root);
+    flock.l_owner.len = frame->root->lk_owner.len;
+    if (flock.l_owner.len > 0) {
+        memcpy(flock.l_owner.data, frame->root->lk_owner.data,
+               flock.l_owner.len);
+    }
+
+    AFR_ONLIST(local->child_up, frame, afr_strict_lock_getlk_cbk, lk, local->fd,
+               F_GETLK, &flock, xdata);
+
+    success_replies = alloca0(priv->child_count);
+    afr_fill_success_replies(local, priv, success_replies);
+    if (AFR_COUNT(success_replies, priv->child_count) == 0)
+        goto out;
+
+    if (!local->strict_locks)
+        goto reopen_fd;
+
+    for (i = 0; i < priv->child_count; i++) {
+        if (!local->replies[i].valid || local->replies[i].op_ret != 0) {
+            continue;
+        }
+        if (local->cont.lk.getlk_rsp1[i].l_type == F_UNLCK) {
+            continue;
+        } else if (local->cont.lk.getlk_rsp1[i].l_type == F_WRLCK) {
+            locks_held = _gf_true;
+            break;
+        }
+    }
+
+    if (locks_held) {
+        gf_log(this->name, GF_LOG_INFO,
+               "client.strict-locks is enabled and conflicting locks are held. "
+               "Skipping fd reopen");
+        goto out;
+    }
+
+reopen_fd:
+    need_open = alloca0(priv->child_count);
+
+    need_open_count = afr_fd_ctx_need_open(local->fd, this, need_open);
+    if (!need_open_count) {
+        goto out;
+    }
+    gf_msg_debug(this->name, 0, "need open count: %d", need_open_count);
+
+    for (i = 0; i < priv->child_count; i++) {
+        if (need_open[i])
+            local->fd_ctx->opened_on[i] = AFR_FD_OPENING;
+    }
+
+    if (IA_IFDIR == local->fd->inode->ia_type) {
+        AFR_ONLIST(need_open, frame, afr_openfd_fix_open_cbk, opendir,
+                   &local->loc, local->fd, NULL);
+    } else {
+        AFR_ONLIST(need_open, frame, afr_openfd_fix_open_cbk, open, &local->loc,
+                   local->fd_ctx->flags & (~O_TRUNC), local->fd, NULL);
+    }
+
+out:
+    GF_FREE(local->cont.lk.getlk_rsp1);
+    return 0;
+}
+
```

Comment:
I will review the function above once you resubmit after converting it back to without synctask.

Reply:
@pranithk I have made all the changes. Please do review.